packages/flutter_tools/lib/src/cache.dart

```diff
@@ -96,7 +96,41 @@ class Cache {
   final List<CachedArtifact> _artifacts = <CachedArtifact>[];
 
   // Initialized by FlutterCommandRunner on startup.
-  static String flutterRoot;
+  static String get flutterRoot => _flutterRoot;
+  static String _flutterRoot;
+  static set flutterRoot(String value) {
+    if (value == null) {
+      _flutterRoot = null;
+      return;
+    }
+    // Verify that we have writable permission in the flutter root. If not,
+    // we're liable to crash in unintuitive ways. This can happen if the user
+    // is using a homebrew or other unofficial channel, or otherwise installs
+    // Flutter into directory without permissions.
+    try {
+      // Check bin.
+      fs.file(fs.path.join(value, 'bin', '.check'))
```

Comment:
nit: I saw we could also use https://api.dartlang.org/stable/2.3.2/dart-io/FileStat/modeString.html from `Directory.statSync`.